File: python_market_data_collector/config_loader.py
"""
Config Loader - Python equivalent of config_loader.cpp

This module provides configuration loading from JSON files,
equivalent to the C++ config_loader.cpp.
"""

import json
import os
from typing import Optional

from .config_types import (
    MarketDataCollectorConfig,
    ExchangeConfig,
    DebugConfig,
    WarmupConfig
)
from .utilities import get_current_timestamp, create_connection_string


def load_config(path: str) -> MarketDataCollectorConfig:
    """
    Load configuration from JSON file or create default configuration

    Args:
        path: Path to JSON configuration file

    Returns:
        MarketDataCollectorConfig object
    """
    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Starting configuration loading from {path}")

    config = MarketDataCollectorConfig()

    if not os.path.exists(path):
        # Create default configuration
        print(f"[{get_current_timestamp()}][WARNING] ConfigLoader: Config file not found: {path}, using default configuration.")

        # Default database connection
        config.db_connection_string = create_connection_string(
            "localhost",
            "market_data",
            "sa",
            "your_password"
        )

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default database connection string created")

        # Default timeframes
        config.timeframes = ["1m", "5m", "15m", "1h"]
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default timeframes set: {', '.join(config.timeframes)}")

        # Default exchange configuration
        ec = ExchangeConfig(
            exchange_name="binance",
            symbols=["BTC-USDT", "ETH-USDT"],
            channels=["TRADE", "MARKET_DEPTH"],
            market_type="spot"
        )
        config.exchanges.append(ec)

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default exchange configuration added: {ec.exchange_name} with {ec.symbols}")

        # Default buffer sizes and intervals
        config.trade_buffer_size = 1000
        config.candle_buffer_size = 200
        config.orderbook_buffer_size = 100
        config.flush_interval_ms = 1000
        config.stats_report_interval_s = 10

        # Default parallel settings
        config.num_worker_threads = 16
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default parallel settings set - num_worker_threads: {config.num_worker_threads}")

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default buffer sizes set:")
        print(f"[{get_current_timestamp()}][INFO]   Trade buffer: {config.trade_buffer_size}")
        print(f"[{get_current_timestamp()}][INFO]   Candle buffer: {config.candle_buffer_size}")
        print(f"[{get_current_timestamp()}][INFO]   Orderbook buffer: {config.orderbook_buffer_size}")
        print(f"[{get_current_timestamp()}][INFO]   Flush interval: {config.flush_interval_ms}ms")
        print(f"[{get_current_timestamp()}][INFO]   Stats interval: {config.stats_report_interval_s}s")

        # Default toggle settings
        config.enable_mssql = True
        config.enable_exclusive_hotspine = False

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default toggle settings:")
        print(f"[{get_current_timestamp()}][INFO]   MS SQL enabled: {config.enable_mssql}")
        print(f"[{get_current_timestamp()}][INFO]   Exclusive HotSpine: {config.enable_exclusive_hotspine}")

        # Default warmup configuration
        config.warmup_config = WarmupConfig(
            batch_size=20,
            startup_delay_ms=5000,
            wait_for_connection=True
        )

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default warmup configuration:")
        print(f"[{get_current_timestamp()}][INFO]   Batch size: {config.warmup_config.batch_size}")
        print(f"[{get_current_timestamp()}][INFO]   Startup delay: {config.warmup_config.startup_delay_ms}ms")
        print(f"[{get_current_timestamp()}][INFO]   Wait for connection: {config.warmup_config.wait_for_connection}")

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default configuration loading completed")

        return config

    # Load configuration from file
    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Loading JSON configuration from file")
    with open(path, 'r') as f:
        j = json.load(f)

    # Load database configuration
    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Loading database configuration")
    db = j.get("db", {})
    server = db.get("server", "localhost")
    database = db.get("database", "market_data")
    user = db.get("user", "sa")
    password = db.get("password", "")

    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Database connection details:")
    print(f"[{get_current_timestamp()}][INFO]   Server: {server}")
    print(f"[{get_current_timestamp()}][INFO]   Database: {database}")
    print(f"[{get_current_timestamp()}][INFO]   User: {user}")
    print(f"[{get_current_timestamp()}][INFO]   Password: [REDACTED]")

    config.db_connection_string = create_connection_string(server, database, user, password)
    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Database connection string created")

    # Load timeframes
    config.timeframes = j.get("timeframes", ["1m", "5m", "15m", "1h"])
    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Timeframes loaded: {', '.join(config.timeframes)}")

    # Load exchange configurations
    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Loading exchange configurations")
    for ex in j.get("exchanges", []):
        ec = ExchangeConfig(
            exchange_name=ex.get("name", ""),
            symbols=ex.get("symbols", []),
            channels=ex.get("channels", ["TRADE", "MARKET_DEPTH"]),
            market_type=ex.get("market_type", "spot")
        )

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Exchange configuration loaded:")
        print(f"[{get_current_timestamp()}][INFO]   Exchange: {ec.exchange_name}")
        print(f"[{get_current_timestamp()}][INFO]   Symbols: {', '.join(ec.symbols)}")
        print(f"[{get_current_timestamp()}][INFO]   Channels: {', '.join(ec.channels)}")
        print(f"[{get_current_timestamp()}][INFO]   Market type: {ec.market_type}")

        config.exchanges.append(ec)

    # Load buffer sizes and intervals
    config.trade_buffer_size = j.get("trade_buffer_size", 500)
    config.candle_buffer_size = j.get("candle_buffer_size", 200)
    config.orderbook_buffer_size = j.get("orderbook_buffer_size", 100)
    config.flush_interval_ms = j.get("flush_interval_ms", 1000)
    config.stats_report_interval_s = j.get("stats_report_interval_s", 10)

    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Buffer and interval settings loaded:")
    print(f"[{get_current_timestamp()}][INFO]   Trade buffer size: {config.trade_buffer_size}")
    print(f"[{get_current_timestamp()}][INFO]   Candle buffer size: {config.candle_buffer_size}")
    print(f"[{get_current_timestamp()}][INFO]   Orderbook buffer size: {config.orderbook_buffer_size}")
    print(f"[{get_current_timestamp()}][INFO]   Flush interval: {config.flush_interval_ms}ms")
    print(f"[{get_current_timestamp()}][INFO]   Stats report interval: {config.stats_report_interval_s}s")

    # Load feature toggles
    config.enable_mssql = j.get("enable_mssql", True)
    config.enable_exclusive_hotspine = j.get("enable_exclusive_hotspine", False)

    # Load parallel settings
    if "parallel_settings" in j:
        parallel = j["parallel_settings"]
        config.num_worker_threads = parallel.get("num_worker_threads", 16)
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Parallel settings loaded - num_worker_threads: {config.num_worker_threads}")
    else:
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: No parallel_settings found, using default (16 threads)")

    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Feature toggle settings loaded:")
    print(f"[{get_current_timestamp()}][INFO]   Enable MS SQL: {config.enable_mssql}")
    print(f"[{get_current_timestamp()}][INFO]   Enable exclusive HotSpine: {config.enable_exclusive_hotspine}")

    # Load debug mode configuration
    if "debug_mode" in j:
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Loading debug mode configuration")
        debug = j["debug_mode"]

        config.debug_config = DebugConfig(
            enabled=debug.get("enabled", False),
            verbose_logging=debug.get("verbose_logging", False),
            websocket_debug=debug.get("websocket_debug", False),
            database_debug=debug.get("database_debug", False),
            buffer_debug=debug.get("buffer_debug", False),
            flush_debug=debug.get("flush_debug", False),
            performance_monitoring=debug.get("performance_monitoring", False),
            connection_health_checks=debug.get("connection_health_checks", False),
            data_flow_validation=debug.get("data_flow_validation", False),
            error_tracing=debug.get("error_tracing", False),
            timestamp_precision=debug.get("timestamp_precision", "milliseconds"),
            log_level=debug.get("log_level", "INFO")
        )

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Debug mode settings loaded:")
        print(f"[{get_current_timestamp()}][INFO]   Debug enabled: {config.debug_config.enabled}")
        print(f"[{get_current_timestamp()}][INFO]   Verbose logging: {config.debug_config.verbose_logging}")
        print(f"[{get_current_timestamp()}][INFO]   WebSocket debug: {config.debug_config.websocket_debug}")
        print(f"[{get_current_timestamp()}][INFO]   Database debug: {config.debug_config.database_debug}")
        print(f"[{get_current_timestamp()}][INFO]   Buffer debug: {config.debug_config.buffer_debug}")
        print(f"[{get_current_timestamp()}][INFO]   Log level: {config.debug_config.log_level}")

        # Load database debug configuration
        if "database_debug" in debug and isinstance(debug["database_debug"], dict):
            db_debug = debug["database_debug"]
            config.debug_config.database_debug_config = DebugConfig.DatabaseDebugConfig(
                verify_connection_on_start=db_debug.get("verify_connection_on_start", False),
                test_table_creation=db_debug.get("test_table_creation", False),
                log_all_queries=db_debug.get("log_all_queries", False),
                log_connection_status=db_debug.get("log_connection_status", False),
                enable_detailed_error_logging=db_debug.get("enable_detailed_error_logging", False),
                connection_timeout_ms=db_debug.get("connection_timeout_ms", 30000),
                command_timeout_ms=db_debug.get("command_timeout_ms", 60000)
            )

        # Load WebSocket debug configuration
        if "websocket_debug" in debug and isinstance(debug["websocket_debug"], dict):
            ws_debug = debug["websocket_debug"]
            config.debug_config.websocket_debug_config = DebugConfig.WebSocketDebugConfig(
                log_connection_attempts=ws_debug.get("log_connection_attempts", False),
                log_subscription_status=ws_debug.get("log_subscription_status", False),
                log_data_reception=ws_debug.get("log_data_reception", False),
                ping_pong_monitoring=ws_debug.get("ping_pong_monitoring", False),
                connection_health_monitoring=ws_debug.get("connection_health_monitoring", False),
                timeout_debugging=ws_debug.get("timeout_debugging", False),
                protocol_error_logging=ws_debug.get("protocol_error_logging", False),
                data_rate_monitoring=ws_debug.get("data_rate_monitoring", False)
            )

        # Load buffer debug configuration
        if "buffer_debug" in debug and isinstance(debug["buffer_debug"], dict):
            buf_debug = debug["buffer_debug"]
            config.debug_config.buffer_debug_config = DebugConfig.BufferDebugConfig(
                log_buffer_operations=buf_debug.get("log_buffer_operations", False),
                log_threshold_crossing=buf_debug.get("log_threshold_crossing", False),
                log_flush_decisions=buf_debug.get("log_flush_decisions", False),
                monitor_memory_usage=buf_debug.get("monitor_memory_usage", False),
                track_buffer_sizes=buf_debug.get("track_buffer_sizes", False),
                log_batch_operations=buf_debug.get("log_batch_operations", False)
            )

        # Load performance monitoring configuration
        if "performance_monitoring" in debug and isinstance(debug["performance_monitoring"], dict):
            perf_debug = debug["performance_monitoring"]
            config.debug_config.performance_config = DebugConfig.PerformanceConfig(
                measure_processing_latency=perf_debug.get("measure_processing_latency", False),
                track_throughput=perf_debug.get("track_throughput", False),
                monitor_resource_usage=perf_debug.get("monitor_resource_usage", False),
                log_performance_metrics=perf_debug.get("log_performance_metrics", False),
                alert_on_slow_operations=perf_debug.get("alert_on_slow_operations", False)
            )

        # If debug mode is disabled, override all sub-flags to false
        if not config.debug_config.enabled:
            config.debug_config.verbose_logging = False
            config.debug_config.websocket_debug = False
            config.debug_config.database_debug = False
            config.debug_config.buffer_debug = False
            config.debug_config.flush_debug = False
            config.debug_config.performance_monitoring = False
            config.debug_config.connection_health_checks = False
            config.debug_config.data_flow_validation = False
            config.debug_config.error_tracing = False

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Debug mode configuration completed successfully")
    else:
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: No debug mode configuration found, using defaults")

    # Load warmup configuration
    if "warmup_config" in j:
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Loading warmup configuration")
        warmup = j["warmup_config"]

        config.warmup_config = WarmupConfig(
            batch_size=warmup.get("batch_size", 20),
            startup_delay_ms=warmup.get("startup_delay_ms", 5000),
            wait_for_connection=warmup.get("wait_for_connection", True)
        )

        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Warmup configuration loaded:")
        print(f"[{get_current_timestamp()}][INFO]   Batch size: {config.warmup_config.batch_size}")
        print(f"[{get_current_timestamp()}][INFO]   Startup delay: {config.warmup_config.startup_delay_ms}ms")
        print(f"[{get_current_timestamp()}][INFO]   Wait for connection: {config.warmup_config.wait_for_connection}")
    else:
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: No warmup_config found, using defaults")
        print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Default warmup configuration:")
        print(f"[{get_current_timestamp()}][INFO]   Batch size: {config.warmup_config.batch_size}")
        print(f"[{get_current_timestamp()}][INFO]   Startup delay: {config.warmup_config.startup_delay_ms}ms")
        print(f"[{get_current_timestamp()}][INFO]   Wait for connection: {config.warmup_config.wait_for_connection}")

    print(f"[{get_current_timestamp()}][INFO] ConfigLoader: Configuration loading completed successfully")

    return config

File: .ralphy-sandboxes/agent-10-a9l4na/dependencies/ccapi/example/src/market_data_collector/config_loader.cpp
#include <fstream>
#include <iostream>
#include "market_data_collector.h"
#include "nlohmann/json.hpp"
#include "utilities.h"

using json = nlohmann::json;

// Helper functions from utilities.h

std::string createConnectionString(const std::string& server,
                                    const std::string& database,
                                    const std::string& user,
                                    const std::string& pwd) {
    return
        "DRIVER={ODBC Driver 18 for SQL Server};"
        "SERVER=" + server + ";"
        "DATABASE=" + database + ";"
        "UID=" + user + ";"
        "PWD=" + pwd + ";"
        "TrustServerCertificate=yes;"
        "MARS_Connection=yes;"
        "Connection Timeout=30;"
        "Command Timeout=60;";
}

MarketDataCollector::Config loadConfig(const std::string& path) {
    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Starting configuration loading from " << path << std::endl;
    
    std::ifstream in(path);
    json j;

    if (!in) {
        // If config file doesnt exist, use default values
        std::cout << "[" << getCurrentTimestamp() << "][WARNING] ConfigLoader: Config file not found: " << path << ", using default configuration." << std::endl;
        
        // Create default configuration
        MarketDataCollector::Config cfg;
        
        // Default database connection (will be used if enable_mssql is true)
        cfg.db_connection_string = createConnectionString(
            "localhost",
            "market_data",
            "sa",
            "your_password");
        
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default database connection string created" << std::endl;
        
        // Default timeframes
        cfg.timeframes = {"1m", "5m", "15m", "1h"};
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default timeframes set: 1m, 5m, 15m, 1h" << std::endl;
        
        // Default exchange configuration
        ExchangeConnectionManager::ExchangeConfig ec;
        ec.exchange_name = "binance";
        ec.symbols = {"BTC-USDT", "ETH-USDT"};
        ec.channels = {"TRADE", "MARKET_DEPTH"};
        ec.market_type = "spot";
        cfg.exchanges.push_back(std::move(ec));
        
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default exchange configuration added: binance with BTC-USDT, ETH-USDT" << std::endl;
        
        // Default buffer sizes and intervals
        cfg.trade_buffer_size = 1000;
        cfg.candle_buffer_size = 200;
        cfg.orderbook_buffer_size = 100;
        cfg.flush_interval_ms = 1000;
        cfg.stats_report_interval_s = 10;

        // Default parallel settings for handling 500+ trading pairs
        cfg.num_worker_threads = 16;
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default parallel settings set - num_worker_threads: " << cfg.num_worker_threads << std::endl;
        
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default buffer sizes set:" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Trade buffer: " << cfg.trade_buffer_size << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Candle buffer: " << cfg.candle_buffer_size << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Orderbook buffer: " << cfg.orderbook_buffer_size << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Flush interval: " << cfg.flush_interval_ms << "ms" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Stats interval: " << cfg.stats_report_interval_s << "s" << std::endl;
        
        // Default toggle settings - MS SQL enabled, hotswap disabled
        cfg.enable_mssql = true;
        cfg.enable_exclusive_hotspine = false;
        
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default toggle settings:" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   MS SQL enabled: " << (cfg.enable_mssql ? "true" : "false") << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Exclusive HotSpine: " << (cfg.enable_exclusive_hotspine ? "true" : "false") << std::endl;
        
        // Default warmup configuration
        cfg.warmup_config.batch_size = 20;
        cfg.warmup_config.startup_delay_ms = 5000;
        cfg.warmup_config.wait_for_connection = true;
        
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default warmup configuration:" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Batch size: " << cfg.warmup_config.batch_size << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Startup delay: " << cfg.warmup_config.startup_delay_ms << "ms" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Wait for connection: " << (cfg.warmup_config.wait_for_connection ? "true" : "false") << std::endl;
        
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default configuration loading completed" << std::endl;
        
        return cfg;
    }
    
    // Load configuration from file
    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Loading JSON configuration from file" << std::endl;
    in >> j;

    MarketDataCollector::Config cfg;

    // Load database configuration
    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Loading database configuration" << std::endl;
    auto db = j.at("db");
    std::string server = db.at("server").get<std::string>();
    std::string database = db.at("database").get<std::string>();
    std::string user = db.at("user").get<std::string>();
    std::string password = db.at("password").get<std::string>();
    
    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Database connection details:" << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Server: " << server << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Database: " << database << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   User: " << user << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Password: [REDACTED]" << std::endl;
    
    cfg.db_connection_string = createConnectionString(server, database, user, password);
    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Database connection string created" << std::endl;

    // Load timeframes
    cfg.timeframes = j.at("timeframes").get<std::vector<std::string>>();
    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Timeframes loaded: ";
    for (const auto& tf : cfg.timeframes) {
        std::cout << tf << " ";
    }
    std::cout << std::endl;

    // Load exchange configurations
    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Loading exchange configurations" << std::endl;
    for (const auto& ex : j.at("exchanges")) {
        ExchangeConnectionManager::ExchangeConfig ec;

        ec.exchange_name = ex.at("name").get<std::string>();
        ec.symbols       = ex.at("symbols")
                             .get<std::vector<std::string>>();
        ec.channels      = ex.value(
            "channels",
            std::vector<std::string>{"TRADE", "MARKET_DEPTH"}
        );
        ec.market_type   = ex.value(
            "market_type",
            std::string("spot")
        );

        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Exchange configuration loaded:" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Exchange: " << ec.exchange_name << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Symbols: ";
        for (const auto& sym : ec.symbols) {
            std::cout << sym << " ";
        }
        std::cout << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Channels: ";
        for (const auto& ch : ec.channels) {
            std::cout << ch << " ";
        }
        std::cout << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Market type: " << ec.market_type << std::endl;

        cfg.exchanges.push_back(std::move(ec));
    }

    // Load buffer sizes and intervals
    cfg.trade_buffer_size        = j.value("trade_buffer_size", 500);
    cfg.candle_buffer_size       = j.value("candle_buffer_size", 200);
    cfg.orderbook_buffer_size    = j.value("orderbook_buffer_size", 100);
    cfg.flush_interval_ms        = j.value("flush_interval_ms", 1000);
    cfg.stats_report_interval_s  = j.value("stats_report_interval_s", 10);

    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Buffer and interval settings loaded:" << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Trade buffer size: " << cfg.trade_buffer_size << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Candle buffer size: " << cfg.candle_buffer_size << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Orderbook buffer size: " << cfg.orderbook_buffer_size << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Flush interval: " << cfg.flush_interval_ms << "ms" << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Stats report interval: " << cfg.stats_report_interval_s << "s" << std::endl;

    // Load new configuration options
    cfg.enable_mssql              = j.contains("enable_mssql") ? j["enable_mssql"].get<bool>() : true;
    cfg.enable_exclusive_hotspine = j.contains("enable_exclusive_hotspine") ? j["enable_exclusive_hotspine"].get<bool>() : false;

    // Load parallel settings
    if (j.contains("parallel_settings")) {
        auto parallel = j.at("parallel_settings");
        cfg.num_worker_threads = parallel.value("num_worker_threads", 16);
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Parallel settings loaded - num_worker_threads: " << cfg.num_worker_threads << std::endl;
    } else {
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: No parallel_settings found, using default (16 threads)" << std::endl;
    }

    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Feature toggle settings loaded:" << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Enable MS SQL: " << (cfg.enable_mssql ? "true" : "false") << std::endl;
    std::cout << "[" << getCurrentTimestamp() << "][INFO]   Enable exclusive HotSpine: " << (cfg.enable_exclusive_hotspine ? "true" : "false") << std::endl;

    // Load debug mode configuration
    if (j.contains("debug_mode")) {
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Loading debug mode configuration" << std::endl;
        auto debug = j.at("debug_mode");
        
        cfg.debug_config.enabled                    = debug.contains("enabled") && debug["enabled"].is_boolean() ? debug["enabled"].get<bool>() : false;
        cfg.debug_config.verbose_logging            = debug.contains("verbose_logging") && debug["verbose_logging"].is_boolean() ? debug["verbose_logging"].get<bool>() : false;
        cfg.debug_config.websocket_debug            = debug.contains("websocket_debug") && debug["websocket_debug"].is_boolean() ? debug["websocket_debug"].get<bool>() : false;
        // Handle database_debug: if boolean, use it; if object or present, enable
        if (debug.contains("database_debug") && debug["database_debug"].is_boolean()) {
            cfg.debug_config.database_debug = debug["database_debug"].get<bool>();
        } else if (debug.contains("database_debug")) {
            cfg.debug_config.database_debug = true;
        }
        cfg.debug_config.buffer_debug               = debug.contains("buffer_debug") && debug["buffer_debug"].is_boolean() ? debug["buffer_debug"].get<bool>() : false;
        cfg.debug_config.flush_debug                = debug.contains("flush_debug") && debug["flush_debug"].is_boolean() ? debug["flush_debug"].get<bool>() : false;
        cfg.debug_config.performance_monitoring     = debug.contains("performance_monitoring") && debug["performance_monitoring"].is_boolean() ? debug["performance_monitoring"].get<bool>() : false;
        cfg.debug_config.connection_health_checks   = debug.contains("connection_health_checks") && debug["connection_health_checks"].is_boolean() ? debug["connection_health_checks"].get<bool>() : false;
        cfg.debug_config.data_flow_validation       = debug.contains("data_flow_validation") && debug["data_flow_validation"].is_boolean() ? debug["data_flow_validation"].get<bool>() : false;
        cfg.debug_config.error_tracing              = debug.contains("error_tracing") && debug["error_tracing"].is_boolean() ? debug["error_tracing"].get<bool>() : false;
        cfg.debug_config.timestamp_precision        = debug.value("timestamp_precision", std::string("milliseconds"));
        cfg.debug_config.log_level                  = debug.value("log_level", std::string("INFO"));
        
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Debug mode settings loaded:" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Debug enabled: " << (cfg.debug_config.enabled ? "true" : "false") << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Verbose logging: " << (cfg.debug_config.verbose_logging ? "true" : "false") << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   WebSocket debug: " << (cfg.debug_config.websocket_debug ? "true" : "false") << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Database debug: " << (cfg.debug_config.database_debug ? "true" : "false") << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Buffer debug: " << (cfg.debug_config.buffer_debug ? "true" : "false") << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Log level: " << cfg.debug_config.log_level << std::endl;
        
        // Load database debug configuration
        if (debug.contains("database_debug") && debug["database_debug"].is_object()) {
            auto dbg = debug.at("database_debug");
            cfg.debug_config.database_debug_config.verify_connection_on_start = dbg.contains("verify_connection_on_start") ? dbg["verify_connection_on_start"].get<bool>() : false;
            cfg.debug_config.database_debug_config.test_table_creation        = dbg.contains("test_table_creation") ? dbg["test_table_creation"].get<bool>() : false;
            cfg.debug_config.database_debug_config.log_all_queries            = dbg.contains("log_all_queries") ? dbg["log_all_queries"].get<bool>() : false;
            cfg.debug_config.database_debug_config.log_connection_status      = dbg.contains("log_connection_status") ? dbg["log_connection_status"].get<bool>() : false;
            cfg.debug_config.database_debug_config.enable_detailed_error_logging = dbg.contains("enable_detailed_error_logging") ? dbg["enable_detailed_error_logging"].get<bool>() : false;
            cfg.debug_config.database_debug_config.connection_timeout_ms      = dbg.value("connection_timeout_ms", 30000);
            cfg.debug_config.database_debug_config.command_timeout_ms         = dbg.value("command_timeout_ms", 60000);
        }
        
        // Load WebSocket debug configuration
        if (debug.contains("websocket_debug") && debug["websocket_debug"].is_object()) {
            auto wsd = debug.at("websocket_debug");
            cfg.debug_config.websocket_debug_config.log_connection_attempts     = wsd.contains("log_connection_attempts") ? wsd["log_connection_attempts"].get<bool>() : false;
            cfg.debug_config.websocket_debug_config.log_subscription_status     = wsd.contains("log_subscription_status") ? wsd["log_subscription_status"].get<bool>() : false;
            cfg.debug_config.websocket_debug_config.log_data_reception          = wsd.contains("log_data_reception") ? wsd["log_data_reception"].get<bool>() : false;
            cfg.debug_config.websocket_debug_config.ping_pong_monitoring        = wsd.contains("ping_pong_monitoring") ? wsd["ping_pong_monitoring"].get<bool>() : false;
            cfg.debug_config.websocket_debug_config.connection_health_monitoring = wsd.contains("connection_health_monitoring") ? wsd["connection_health_monitoring"].get<bool>() : false;
            cfg.debug_config.websocket_debug_config.timeout_debugging            = wsd.contains("timeout_debugging") ? wsd["timeout_debugging"].get<bool>() : false;
            cfg.debug_config.websocket_debug_config.protocol_error_logging       = wsd.contains("protocol_error_logging") ? wsd["protocol_error_logging"].get<bool>() : false;
            cfg.debug_config.websocket_debug_config.data_rate_monitoring         = wsd.contains("data_rate_monitoring") ? wsd["data_rate_monitoring"].get<bool>() : false;
        }
        
        // Load buffer debug configuration
        if (debug.contains("buffer_debug") && debug["buffer_debug"].is_object()) {
            auto bd = debug.at("buffer_debug");
            cfg.debug_config.buffer_debug_config.log_buffer_operations  = bd.contains("log_buffer_operations") ? bd["log_buffer_operations"].get<bool>() : false;
            cfg.debug_config.buffer_debug_config.log_threshold_crossing = bd.contains("log_threshold_crossing") ? bd["log_threshold_crossing"].get<bool>() : false;
            cfg.debug_config.buffer_debug_config.log_flush_decisions    = bd.contains("log_flush_decisions") ? bd["log_flush_decisions"].get<bool>() : false;
            cfg.debug_config.buffer_debug_config.monitor_memory_usage   = bd.contains("monitor_memory_usage") ? bd["monitor_memory_usage"].get<bool>() : false;
            cfg.debug_config.buffer_debug_config.track_buffer_sizes     = bd.contains("track_buffer_sizes") ? bd["track_buffer_sizes"].get<bool>() : false;
            cfg.debug_config.buffer_debug_config.log_batch_operations   = bd.contains("log_batch_operations") ? bd["log_batch_operations"].get<bool>() : false;
        }
        
        // Load performance monitoring configuration
        if (debug.contains("performance_monitoring") && debug["performance_monitoring"].is_object()) {
            auto pm = debug.at("performance_monitoring");
            cfg.debug_config.performance_config.measure_processing_latency = pm.contains("measure_processing_latency") ? pm["measure_processing_latency"].get<bool>() : false;
            cfg.debug_config.performance_config.track_throughput           = pm.contains("track_throughput") ? pm["track_throughput"].get<bool>() : false;
            cfg.debug_config.performance_config.monitor_resource_usage     = pm.contains("monitor_resource_usage") ? pm["monitor_resource_usage"].get<bool>() : false;
            cfg.debug_config.performance_config.log_performance_metrics    = pm.contains("log_performance_metrics") ? pm["log_performance_metrics"].get<bool>() : false;
            cfg.debug_config.performance_config.alert_on_slow_operations   = pm.contains("alert_on_slow_operations") ? pm["alert_on_slow_operations"].get<bool>() : false;
        }
        
        // If debug mode is disabled, override all sub-flags to false
        if (!cfg.debug_config.enabled) {
            cfg.debug_config.verbose_logging = false;
            cfg.debug_config.websocket_debug = false;
            cfg.debug_config.database_debug = false;
            cfg.debug_config.buffer_debug = false;
            cfg.debug_config.flush_debug = false;
            cfg.debug_config.performance_monitoring = false;
            cfg.debug_config.connection_health_checks = false;
            cfg.debug_config.data_flow_validation = false;
            cfg.debug_config.error_tracing = false;
        }

        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Debug mode configuration completed successfully" << std::endl;
    } else {
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: No debug mode configuration found, using defaults" << std::endl;
    }

    // Load warmup configuration
    if (j.contains("warmup_config")) {
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Loading warmup configuration" << std::endl;
        auto warmup = j.at("warmup_config");
        
        cfg.warmup_config.batch_size = warmup.value("batch_size", 20);
        cfg.warmup_config.startup_delay_ms = warmup.value("startup_delay_ms", 5000);
        cfg.warmup_config.wait_for_connection = warmup.value("wait_for_connection", true);
        
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Warmup configuration loaded:" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Batch size: " << cfg.warmup_config.batch_size << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Startup delay: " << cfg.warmup_config.startup_delay_ms << "ms" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Wait for connection: " << (cfg.warmup_config.wait_for_connection ? "true" : "false") << std::endl;
    } else {
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: No warmup_config found, using defaults" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Default warmup configuration:" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Batch size: " << cfg.warmup_config.batch_size << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Startup delay: " << cfg.warmup_config.startup_delay_ms << "ms" << std::endl;
        std::cout << "[" << getCurrentTimestamp() << "][INFO]   Wait for connection: " << (cfg.warmup_config.wait_for_connection ? "true" : "false") << std::endl;
    }

    std::cout << "[" << getCurrentTimestamp() << "][INFO] ConfigLoader: Configuration loading completed successfully" << std::endl;

    return cfg;
}